### DISCUSSION TITLE
Add an optional 'from' property to next/router and automatically populate this with current the pathname on router.push()

### DISCUSSION THREAD

**Discussion by U-4-E-A** [vercel/next.js#28184] [Ideas]
### Describe the feature you'd like to request

Currently, there is no intrinsic way to populate the router properties with information regarding from which path `router.push()` was called [like you can do in react-router](https://reactrouter.com/web/api/Redirect/from-string).

Example use case - for when an unauthenticated user is directed to `/login` from `/my-account` and the app then has the option to gracefully redirect the user back to that page after the completion of successful authentication. 

### Describe the solution you'd like

I would like next/router to have an optional `from` property that will be automatically populated when `router.push()` is called.

### Describe alternatives you've considered

Could extend on own codebase but it would seem universally useful.

**Comment by vishal-rajole** [comment]
I am also looking for similar approach, any solution?

**Reply by U-4-E-A** [reply]
Not yet, I have been working on other things. I need to get back to it. I can't see why it hasn't been included as a feature though...

**Reply by U-4-E-A** [reply]
I am thinking about just adding a property to my Apollo local state with a redirection path and have the login form query for the value, redirecting if the value is not `null` then setting the value to `null` regardless when the component dismounts. Not my preferred solution though and feels very hackish.

**Reply by U-4-E-A** [reply]
In the end, I resolved it by using local state. I have an authenticated route component that wraps certain pages. It checks to see if the user is authenticated and redirects to login if not. I have a hook in the conditional that does the redirecting that takes a callback. In this case, the callback I give it updates my Apollo local state with the current route. In my login component, after a successful login, the component queries the local state to see if there is a redirection path. If so, it redirects to it. In the login component's `componentWillUnmount()` there is a local state mutation that sets the redirection path to `null`.

**Comment by lukebussey** [comment]
I use a `usePrevious()` hook for this:

```js
import { useRef, useEffect } from 'react';

const usePrevious = (value) => {
  // The ref object is a generic container whose current property is mutable ...
  // ... and can hold any value, similar to an instance property on a class
  const ref = useRef();

  // Store current value in ref
  useEffect(() => {
    ref.current = value;
  }, [value]); // Only re-run if value changes

  // Return previous value (happens before update in useEffect above)
  return ref.current;
};

export default usePrevious;
```

```js
  const [location, setLocation] = useState();
  const router = useRouter();

  const referrer = usePrevious(location);

  useEffect(() => {
    setLocation(window.location.href);
  }, [router.asPath]);
```

**Reply by peteristhegreat** [reply]
This looks really interesting.  I am pretty new to nextjs.  Could you show it in the context of login pattern?

Here is my existing login pattern:

```
import { useAuth } from 'hooks/useAuth';
import { useEffect } from 'react';
import { useRouter } from 'next/router';

export const useRequireAuth = (redirectUrl = '/login') => {
  const auth = useAuth();
  const router = useRouter();
  // If auth.user is false that means we're not
  // logged in and should redirect.
  useEffect(() => {
    if (!auth.user) {
      router.push({
        pathname: redirectUrl,
        query: { from: router.pathname }, // stores the "from" url into the query
      });
    }
  }, [auth, redirectUrl, router]);

  return auth;
};
```
Then in the other side of the login form (after the login is successful), I have this `useEffect` watching the auth state:
```
  useEffect(() => {
    if (user) {
      if (router.query && router.query.from) {
        router.push(router.query.from as string);  // gets "from" url in the query and routes to "from"
      } else {
        router.push('/dashboard');
      }
    }
  }, [user, router]);
```

Is `useRef` persistent across those pages and can be used just like I use the `query.from` like I do above?  Can you show me where you would initialize the value and retrieve the value in the context of my example?  Does it work with `router.push`, or is it better with `router.asPath`?
